itwm_example/tests/gradient_consistency/taylor_convergence.py

```diff
@@ -0,0 +1,140 @@
+        return regression.slope
+
+    def _test_directions(self):
+        """ Generates simple complete set of the test directions.
+
+        Returns
+        -------
+        directions: list of np.array
+            List of all single component vectors as test directions.
+        """
+        directions = [
+            self._single_component_vector(i)
+            for i in range(self._input_dimension)
+        ]
+        return directions
+
+    def _generate_uniform_perturbations(self, step_size, length=None):
+        if length is None:
+            length = self._default_nof_evaluations
+        perturbations = (
+            (i + 1) * step_size for i in range(length)
+        )
+        return perturbations
+
+    def _calculate_taylor_remainders(
+            self,
+            init_point,
+            direction,
+            step_size=None
+    ):
+        """ Runs a series of function evaluation at the given point,
+        perturbing the evaluation point in the provided direction.
+
+        Returns
+        -------
+        shifts: np.array
+            Array of perturbation magnitude (shifts) used to produce
+            perturbed function values
+        taylor_remainders: np.array
+            Array of calculated Taylor remainders
+        """
+        if step_size is None:
+            step_size = self._default_step_size
+
+        default_value = self._evaluate_function(init_point)
+        perturbation_norm = np.dot(
+            direction,
+            self._evaluate_gradient(init_point)
+        )
+
+        taylor_remainders = np.zeros(self._default_nof_evaluations)
+        shifts = np.zeros(self._default_nof_evaluations)
+        for i, shift in enumerate(
+                self._generate_uniform_perturbations(step_size)
+        ):
+            perturbed_point = init_point.copy()
+            perturbed_point += shift * direction
+
+            perturbed_value = self._evaluate_function(perturbed_point)
+            taylor_remainders[i] = abs(
+                    perturbed_value
+                    - default_value
+                    - shift * perturbation_norm
+            )
+            shifts[i] = shift
+
+        return shifts, taylor_remainders
+
+    def run_taylor_test(self, initial_point):
+        """An entry point to the Taylor testing.
+        Estimates the slope of the Taylor reminders power fit
+        for a complete set of possible directions.
+
+        Returns
+        -------
+        slopes: np.array
+            Array of slopes with respect to the perturbations
+            in the unit directions.
+        """
+        initial_point = np.array(initial_point)
+
+        slopes = np.zeros(self._input_dimension)
+        for i, direction in enumerate(self._test_directions()):
+            x, y = self._calculate_taylor_remainders(
+                initial_point,
+                direction
+            )
+            slopes[i] = self._fit_power_law(x, y)
+
+        return slopes
```

Comment:
Maybe we can return a boolean that indicates whether the test has passed (ie. - what we expect the outcome of slopes to be), or rename the function to `get_slopes` and include the actual testing in another function.

Ideally it would be useful to the api to have something like:
    self.assertTrue(taylor_test.run_taylor_test(initial_points))

Rather than:
    slopes = taylor_test.run_taylor_tests(initial_points)
    for slope in slopes:
        self.assertAlmostEqual(2, slope, 1)

Reply:
I was thinking if we could merge this PR to start implementing the gradient tests for `DataSource`'s. Please let me know what you think.